src/style-spec/expression/definitions/collator.js

```diff
@@ -0,0 +1,141 @@
+  caseFirst?: 'upper' | 'lower' | 'false'
+}
+
+export class CollatorInstantiation {
+    locale: string | null;
+    sensitivity: 'base' | 'accent' | 'case' | 'variant';
+
+    constructor(caseSensitive: boolean, diacriticSensitive: boolean, locale: string | null) {
+        if (caseSensitive)
+            this.sensitivity = diacriticSensitive ? 'variant' : 'case';
+        else
+            this.sensitivity = diacriticSensitive ? 'accent' : 'base';
+
+        this.locale = locale;
+    }
+
+    compare(lhs: string, rhs: string): number {
+        return new Intl.Collator(this.locale ? this.locale : [],
+                                 { sensitivity: this.sensitivity, usage: 'search' })
+            .compare(lhs, rhs);
+    }
+
+    resolvedLocale(): string {
+        return new Intl.Collator(this.locale ? this.locale : [])
+            .resolvedOptions().locale;
+    }
+
+    serialize() {
+        const options = {};
+        options['caseSensitive'] = this.sensitivity === 'variant' || this.sensitivity === 'case';
+        options['diacriticSensitive'] = this.sensitivity === 'variant' || this.sensitivity === 'accent';
+        if (this.locale) {
+            options['locale'] = this.locale;
+        }
+        return ["collator", options];
+    }
+}
+
+export class CollatorExpression implements Expression {
+    type: Type;
+    caseSensitive: Expression;
+    diacriticSensitive: Expression;
+    locale: Expression | null;
+
+    constructor(caseSensitive: Expression, diacriticSensitive: Expression, locale: Expression | null) {
+        this.type = CollatorType;
+        this.locale = locale;
+        this.caseSensitive = caseSensitive;
+        this.diacriticSensitive = diacriticSensitive;
+    }
+
+    static parse(args: Array<mixed>, context: ParsingContext): ?Expression {
+        if (args.length !== 2)
+            return context.error(`Expected one argument.`);
+
+        const options = (args[1]: any);
+        if (typeof options !== "object" || Array.isArray(options))
+            return context.error(`Collator options argument must be an object.`);
+
+        const caseSensitive = context.parse(
+            options['caseSensitive'] === undefined ? false : options['caseSensitive'], 1, BooleanType);
+        if (!caseSensitive) return null;
+
+        const diacriticSensitive = context.parse(
+            options['diacriticSensitive'] === undefined ? false : options['diacriticSensitive'], 2, BooleanType);
+        if (!diacriticSensitive) return null;
+
+        let locale = null;
+        if (options['locale']) {
+            locale = context.parse(options['locale'], 3, StringType);
+            if (!locale) return null;
+        }
+
+        return new CollatorExpression(caseSensitive, diacriticSensitive, locale);
+    }
+
+    evaluate(ctx: EvaluationContext) {
+        return new CollatorInstantiation(this.caseSensitive.evaluate(ctx), this.diacriticSensitive.evaluate(ctx), this.locale ? this.locale.evaluate(ctx) : null);
```

Comment:
You mean the cost of `new Intl.Collator` (which we do on each `compare`)? It would be easy enough to move into the `CollatorInstantiation` constructor (allowing all of the uses in one expression to share the same constructor). But doesn't seem like something we should bother doing unless we see it showing up as a hot spot...